File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/base/BaseActivity.java
package com.jiujiu.mvvmTemplate.ui.base;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.LayoutRes;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import javax.inject.Inject;

import dagger.android.AndroidInjection;
import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.support.DaggerAppCompatActivity;
import dagger.android.support.HasSupportFragmentInjector;

public abstract class BaseActivity<T extends ViewDataBinding, V extends ViewModel> extends DaggerAppCompatActivity {

    @Inject
    ViewModelProvider.Factory factory;

    private T mBinding;
    private V viewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        performViewModel();
        performDataBinding();
    }

    private void performViewModel() {
        viewModel = ViewModelProviders.of(this, factory).get(getViewModelClassType());
    }

    protected abstract Class<V> getViewModelClassType();

    @LayoutRes
    public abstract int getLayoutId();

    public V getViewModel() {
        return viewModel;
    }

    public void performDataBinding() {
        mBinding = DataBindingUtil.setContentView(this, getLayoutId());
        mBinding.setLifecycleOwner(this);
        viewModel = viewModel == null ? getViewModel() : viewModel;
        mBinding.setVariable(getBindingVariableId(), viewModel);
        mBinding.executePendingBindings();
    }

    public abstract int getBindingVariableId();

    public T getBinding() {
        return mBinding;
    }

    public void hideKeyboard() {
        View view = this.getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            if (imm != null) {
                imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
            }
        }
    }

}

File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/base/BaseActivity.kt
package com.jiujiu.mvvmTemplate.ui.base

import android.content.Context
import android.os.Bundle
import android.view.inputmethod.InputMethodManager
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import dagger.android.support.DaggerAppCompatActivity
import org.jetbrains.anko.AnkoLogger
import javax.inject.Inject

abstract class BaseActivity<T : ViewDataBinding, V : ViewModel> : DaggerAppCompatActivity(), AnkoLogger {

    @Inject
    lateinit var factory: ViewModelProvider.Factory

    lateinit var binding: T
        private set

    lateinit var viewModel: V
        private set

    protected abstract val viewModelClassType: Class<V>

    @get:LayoutRes
    abstract val layoutId: Int

    abstract val bindingVariableId: Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        performViewModel()
        performDataBinding()
    }

    private fun performViewModel() {
        viewModel = ViewModelProviders.of(this, factory).get(viewModelClassType)
    }

    private fun performDataBinding() {
        binding = DataBindingUtil.setContentView(this, layoutId)
        binding.lifecycleOwner = this
        binding.setVariable(bindingVariableId, viewModel)
        binding.executePendingBindings()
    }

    fun hideKeyboard() {
        val view = this.currentFocus
        if (view != null) {
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

}
